File: arrays_and_strings/1-1.js
// Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

// Uses additional data structures. I am unsure of efficiency.
function uniqueChars(str) {
  var filtered = str.split("").filter(function(element, index, array){
    return array.lastIndexOf(element) === index
  }).join("")
  return filtered.length === str.length
}

console.log(uniqueChars("alalalala"))
console.log(uniqueChars("WEWLAD"))
console.log(uniqueChars("aylmob"))

File: arrays_and_strings/1-1.rb
# Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

# Not written

# def is_unique?(string)
#   return string.split("").uniq.length == string.split("").length
# end

# O(n^2) I believe
# def is_unique?(string)
#   length = string.length
#   for x in 1...length
#     for y in 1...length
#       return false if string[x] == string[y] && x != y
#     end
#   end
#   return true
# end

def is_unique?(string)
  for x in 1...string.length
    for y in (x+1)...string.length
      return false if string[x] == string[y]
    end
  end
  return true
end

p is_unique?("lalalala")
p is_unique?("ay chilbro")